src/Workspaces/Remote/ServiceHub/Host/SolutionAssetSource.cs

```diff
@@ -31,10 +32,22 @@ public async ValueTask<ImmutableArray<(Checksum, object)>> GetAssetsAsync(Checks
             using var provider = await _client.GetProxyAsync<ISolutionAssetProvider>(SolutionAssetProvider.ServiceDescriptor, cancellationToken).ConfigureAwait(false);
             Contract.ThrowIfNull(provider.Proxy);
 
-            return await new RemoteCallback<ISolutionAssetProvider>(provider.Proxy).InvokeAsync(
-                (proxy, pipeWriter, cancellationToken) => proxy.GetAssetsAsync(pipeWriter, solutionChecksum, checksums.ToArray(), cancellationToken),
+            var optional = await new RemoteCallback<ISolutionAssetProvider>(provider.Proxy).InvokeAsync(
+                (proxy, pipeWriter, cancellationToken) => proxy.TryGetAssetsAsync(pipeWriter, solutionChecksum, checksums.ToArray(), cancellationToken),
                 (pipeReader, cancellationToken) => RemoteHostAssetSerialization.ReadDataAsync(pipeReader, solutionChecksum, checksums, serializerService, cancellationToken),
                 cancellationToken).ConfigureAwait(false);
+
+            if (!optional.HasValue)
+                throw new AssetsNotAvailableException();
```

Comment:
i don't love this.  Seeing if i can bubble things up with an optional/nullable without a ton of pain.